File: AlgorithmsPackPythonSolutions/Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.permute_bt(nums, 0, result)
        return result

    def permute_bt(self, nums, index, result):
        if index == len(nums):
            result.append(nums[:])
            return

        for i in range(index, len(nums)):
            nums[index], nums[i] = nums[i], nums[index]
            self.permute_bt(nums, index + 1, result)
            nums[index], nums[i] = nums[i], nums[index]

File: AlgorithmsPackCPP/Permutations.cpp
vector<vector<int>>result;
vector<int>nums;

void permute_bt(int index){
    if(index==nums.size()){
        result.push_back(nums);
        return;
    }

    for(int i = index; i < nums.size(); i++){
        swap(nums[i],nums[index]);
        permute_bt(index+1);
        swap(nums[i],nums[index]);
    }
}

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums_) {
        nums = nums_;
        result.clear();
        permute_bt(0);
        return result;
    }
};

/*
We will permute all iterations/permutations for each index...
swapping the numbers each time
*/
